### PR TITLE
Fix conch and cftp under Python 3

### PR DESCRIPTION
**Pull request by rodrigc**
http://twistedmatrix.com/trac/ticket/8791

### CODE REVIEW

File: src/twisted/conch/scripts/cftp.py [hunk 1]
```diff
@@ -181,40 +190,42 @@ def _dispatchCommand(self, line):
         if f is not None:
             return defer.maybeDeferred(f, rest)
         else:
-            self._ebCommand(failure.Failure(NotImplementedError(
-                "No command called `%s'" % command)))
+            errMsg = "No command called `%s'" % (command)
+            self._ebCommand(failure.Failure(NotImplementedError(errMsg)))
```
Comment (acabhishek942) [inline src/twisted/conch/scripts/cftp.py:194]: why is this change necessary?

File: src/twisted/conch/test/test_cftp.py [hunk 2]
```diff
@@ -994,24 +1041,34 @@ def testCdPwd(self):
         remote directory.
         """
         # XXX - not actually a unit test, see docstring.
-        homeDir = os.path.join(os.getcwd(), self.testDir)
+        homeDir = self.testDir
         d = self.runScript('pwd', 'lpwd', 'cd testDirectory', 'cd ..', 'pwd')
-        d.addCallback(lambda xs: xs[:3] + xs[4:])
+
+        def cmdOutput(output):
```
Comment (acabhishek942) [inline src/twisted/conch/test/test_cftp.py:1047]: would be better to add docstring for this method

File: src/twisted/conch/test/test_cftp.py [hunk 3]
```diff
@@ -1304,6 +1371,7 @@ def _cleanup(res):
 
         return d
 
+
     def testBatchFile(self):
         """Test whether batch file function of cftp ('cftp -b batchfile').
```
Comment (acabhishek942) [inline src/twisted/conch/test/test_cftp.py]: this docstring also needs to be cleaned up

File: src/twisted/conch/test/test_cftp.py [hunk 4]
```diff
@@ -1312,31 +1380,35 @@ def testBatchFile(self):
...
+            self.assertEqual(res[3:-2], [b'testDirectory', b'testRemoveFile',
+                                             b'testRenameFile', b'testfile1'])
 
         d = self._getBatchOutput(cmds)
         d.addCallback(_cbCheckResult)
         return d
 
+
     def testError(self):
         """Test that an error in the batch file stops running the batch.
```
Comment (acabhishek942) [inline src/twisted/conch/test/test_cftp.py]: same here

File: src/twisted/conch/test/test_cftp.py [hunk 5]
```diff
@@ -1312,31 +1380,35 @@ def testBatchFile(self):
...
-            self.assertNotIn(self.testDir, res)
+            self.assertNotIn(self.testDir.asBytesMode().path, res)
 
         d = self._getBatchOutput(cmds)
         d.addCallback(_cbCheckResult)
         return d
 
+
     def testIgnoredError(self):
         """Test that a minus sign '-' at the front of a line ignores
```
Comment (acabhishek942) [inline src/twisted/conch/test/test_cftp.py]: same here

File: src/twisted/conch/test/test_conch.py [hunk 6]
```diff
@@ -490,9 +498,9 @@ def test_clientRekey(self):
         process = ConchTestOpenSSHProcess()
         d = self.execute("", process, '-o RekeyLimit=2K')
         def finished(result):
-            self.assertEqual(
-                result,
-                '\n'.join(['line #%02d' % (i,) for i in range(60)]) + '\n')
+            expectedResult = '\n'.join(['line #%02d' % (i,) for i in range(60)]) + '\n'
+            expectedResult = expectedResult.encode("utf-8")
+            self.assertEqual(result, expectedResult)
```
Comment (acabhishek942) [inline src/twisted/conch/test/test_conch.py:503]: I feel there is a redundant line here it could be changed to `self.assertEqual(result, expectedResult.encode('utf-8'))` and remove this line `expectedResult = expectedResult.encode("utf-8")`

**Review by acabhishek942 — CHANGES_REQUESTED** [review]
I have left some inline comments, please address them as you wish, my one other concern is the coverage issue which is bad mainly because of the unicode check performed at several places. I feel this could be improved by writing a helper function in twisted.python.compat or in some other module(I am not exactly sure which module should have this helper method)

File: src/twisted/conch/scripts/cftp.py (hunk 1 above)
Comment (rodrigc) [reply src/twisted/conch/scripts/cftp.py:194]: I separated the variables out to facilitate debugging as I was modifying the code.
